platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -17,68 +19,374 @@ components within the platform.
+lifecycle will be enabled by setting the flag to `true`.
 
-### Roster Public API
+### Architecture and/or Components
 
-#### Setting the Candidate Roster
+No new components are proposed. Without loss of capability, the platform `AddressBook` data structure is replaced in
+all platform components with the `Roster` data structure, including in the `ConsensusRound` data structure that is
+passed to the application. If the application is dependent on any of the removed fields on an address in the
+`AddressBook` data structure, it will have to look that information up from its own data storage for the Hedera Address
+Book.
 
-A new method will be added to the platform state to allow the App to submit a Candidate Roster to the platform:
+Updating the roster is now performed through writing a `candidate` roster in the state instead of writing a new
+`config.txt` and the associated gossip certificates to disk.
 
-```java
-//in PlaformState
+#### Roster Types.
 
-void setCandidateRoster(@NonNull final Roster candidateRoster);
-```
+This proposal introduces four types of `Roster` - `active`, `candidate`, `genesis`, and `override`.
+
+1. `Active Rosters` are paired with the round numbers for when they became active and are used to validate event
+   signatures. The term `Active` refers to their continued relevancy for the system to operate, including restarts. The
+   active roster with the latest round number is used to determine consensus.
+   * The `Roster History` is a list of active rosters paired with the rounds they became active. At this time the
+     history length is at most 2.
+     * `currentRoster` - the active roster in the history paired with the highest round number.
+     * `previousRoster` - the active roster in the history paired with the lowest round number.
+     * In this document we denote the roster history as an ordered list of pairs by descending round
+       number. Example:  [(currentRoster, 2), (previousRoster, 1)].
+2. A `Genesis Roster` is loaded from disk at the genesis of a new network and becomes the first active roster.
+3. The `Candidate Roster` is a roster that will become active at the next software upgrade.
+4. An `Override Roster` is loaded from disk and unconditionally set as the latest active roster. This is used to
+   transplant the state to a different network.  Until the TSS effort is delivered and the network is given a ledger
+   id, this override mechanism can be used to manually correct the roster / address book on the current network.
+
+### Core Behaviors: Roster Lifecycle
+
+The following diagram depicts the new roster life-cycle where the `Roster History` provided to the platform is
+determined at startup and committed to the merkle state before handing it to the platform.
+
+On software upgrade, during normal execution, if a candidate roster is present in the state, it will be unconditionally
```

Comment:
What should be the behavior if the candidate roster is not present in the state?
By "unconditionally" is it meant that we should not check if we can recreate the ledger id from the candidate roster i.e. that has already happened at some earlier moment?

Reply:
This is the TSS-Roster proposal, which should deliver before TSS-Ledger-ID and there is no concept of a ledger id in this proposal.   This was meant to adapt the Dynamic Address Book Phase 2 life-cycle to using rosters.   The config.txt address book from disk is adopted unconditionally at software upgrade boundaries.  

If there is no candidate roster in the state, and its a software upgrade, and there is an active roster in the roster state, then this is an error scenario.  The changes implied by TSS-Roster have not been implemented properly.

Reply:
> If there is no candidate roster in the state, and its a software upgrade, and there is an active roster in the roster state, then this is an error scenario.

Why does this have to be an error? Why cannot a software upgrade happen w/o any changes to the roster? Why cannot the system simply continue to use the existing active roster in this case?

Reply:
Config.txt is written to disk unconditionally.   Similarly a candidate roster should be written to state unconditionally. 

It is easier to debug scenarios when it is known that there will always be a next roster as an artifact than to try to figure out why a next roster was not generated.

Reply:
That seems a bit strict a requirement, but okay.